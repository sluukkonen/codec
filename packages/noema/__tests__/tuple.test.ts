import { dateFromISOString, number, string, tuple } from "../src/index.js"
import { expectParseFailure, expectParseSuccess } from "./helpers.js"

describe("tuple", () => {
  test("should parse a simple tuples", () => {
    expectParseSuccess(tuple([]), [])
    expectParseSuccess(tuple([number]), [1])
    expectParseSuccess(tuple([number, string]), [1, ""])
  })

  test("should parse complex tuples", () => {
    const now = new Date()
    expectParseSuccess(tuple([dateFromISOString]), [now.toISOString()], [now])
  })

  test("should reject arrays that have incorrect length", () => {
    expectParseFailure(
      tuple([number]),
      [],
      [{ code: "invalid_tuple", path: [] }]
    )
    expectParseFailure(
      tuple([number]),
      [1, 2],
      [{ code: "invalid_tuple", path: [] }]
    )
  })

  test("should reject invalid values", () => {
    expectParseFailure(tuple([]), null, [{ code: "invalid_tuple", path: [] }])
    expectParseFailure(
      tuple([number, number]),
      [1, "2"],
      [{ code: "invalid_number", path: [1] }]
    )
  })
})
