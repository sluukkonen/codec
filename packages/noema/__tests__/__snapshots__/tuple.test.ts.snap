// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should reject arrays that have incorrect length 1`] = `
[
  {
    "code": "too_short",
    "length": 0,
    "maxLength": 1,
    "minLength": 1,
    "path": [],
  },
]
`;

exports[`should reject arrays that have incorrect length 2`] = `
[
  {
    "code": "too_long",
    "length": 2,
    "maxLength": 1,
    "minLength": 1,
    "path": [],
  },
]
`;

exports[`should reject invalid values 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "array",
    "path": [],
    "received": "null",
  },
]
`;

exports[`should reject invalid values 2`] = `
[
  {
    "code": "invalid_type",
    "expected": "number",
    "path": [
      1,
    ],
    "received": "string",
  },
]
`;
